crates/circuit/src/converters.rs

```diff
@@ -85,7 +90,43 @@ pub fn circuit_to_dag(
     )
 }
 
+#[pyfunction(signature = (dag, copy_operations = true))]
+pub fn dag_to_circuit(
+    py: Python,
+    dag: &DAGCircuit,
+    copy_operations: bool,
+) -> PyResult<CircuitData> {
+    CircuitData::from_packed_instructions(
+        py,
+        dag.qubits.clone(),
+        dag.clbits.clone(),
+        dag.qargs_interner.clone(),
+        dag.cargs_interner.clone(),
+        dag.topological_op_nodes()?.map(|node_index| {
+            let NodeType::Operation(ref instr) = dag.dag[node_index] else {
+                unreachable!()
+            };
+            if copy_operations {
+                let op = instr.op.py_deepcopy(py, None)?;
+                Ok(PackedInstruction {
+                    op,
+                    qubits: instr.qubits,
+                    clbits: instr.clbits,
+                    params: instr.params.clone(),
+                    extra_attrs: instr.extra_attrs.clone(),
```

Comment:
You could use `clone_ref(py)` with `Param` now. It doesn't make a big difference but it works well.
```suggestion
                    params: instr.params.iter().map(|param| param.clone_ref(py)).collect(),
```